source/extensions/clusters/redis/redis_cluster.cc

```diff
@@ -0,0 +1,320 @@
+    Network::DnsResolverSharedPtr dns_resolver,
+    Server::Configuration::TransportSocketFactoryContext& factory_context,
+    Stats::ScopePtr&& stats_scope, bool added_via_api)
+    : Upstream::BaseDynamicClusterImpl(cluster, runtime, factory_context, std::move(stats_scope),
+                                       added_via_api),
+      cluster_manager_(clusterManager),
+      cluster_refresh_rate_(std::chrono::milliseconds(
+          PROTOBUF_GET_MS_OR_DEFAULT(redisCluster, cluster_refresh_rate, 5000))),
+      cluster_refresh_timeout_(std::chrono::milliseconds(
+          PROTOBUF_GET_MS_OR_DEFAULT(redisCluster, cluster_refresh_timeout, 3000))),
+      dispatcher_(factory_context.dispatcher()), dns_resolver_(std::move(dns_resolver)),
+      dns_lookup_family_(Upstream::getDnsLookupFamilyFromCluster(cluster)),
+      load_assignment_(cluster.has_load_assignment()
+                           ? cluster.load_assignment()
+                           : Config::Utility::translateClusterHosts(cluster.hosts())),
+      local_info_(factory_context.localInfo()), random_(factory_context.random()),
+      redis_discovery_session_(
+          std::make_unique<RedisDiscoverySession>(*this, redis_client_factory)) {
+  const auto& locality_lb_endpoints = load_assignment_.endpoints();
+  for (const auto& locality_lb_endpoint : locality_lb_endpoints) {
+    for (const auto& lb_endpoint : locality_lb_endpoint.lb_endpoints()) {
+      const auto& host = lb_endpoint.endpoint().address();
+      dns_discovery_resolve_targets_.emplace_back(new DnsDiscoveryResolveTarget(
+          *this, host.socket_address().address(), host.socket_address().port_value(),
+          locality_lb_endpoint, lb_endpoint));
+    }
+  }
+};
+
+void RedisCluster::startPreInit() {
+  for (const DnsDiscoveryResolveTargetPtr& target : dns_discovery_resolve_targets_) {
+    target->startResolve();
+  }
+}
+
+void RedisCluster::updateAllHosts(const Upstream::HostVector& hosts_added,
+                                  const Upstream::HostVector& hosts_removed,
+                                  uint32_t current_priority) {
+  Upstream::PriorityStateManager priority_state_manager(*this, local_info_, nullptr);
+
+  auto locality_lb_endpoint = localityLbEndpoint();
+  priority_state_manager.initializePriorityFor(locality_lb_endpoint);
+  for (const Upstream::HostSharedPtr& host : hosts_) {
+    if (locality_lb_endpoint.priority() == current_priority) {
+      priority_state_manager.registerHostForPriority(host, locality_lb_endpoint);
+    }
+  }
+
+  priority_state_manager.updateClusterPrioritySet(
+      current_priority, std::move(priority_state_manager.priorityState()[current_priority].first),
+      hosts_added, hosts_removed, absl::nullopt);
+}
+
+void RedisCluster::onClusterSlotUpdate(const std::vector<ClusterSlot>& slots) {
+  Upstream::HostVector new_hosts;
+  std::array<Upstream::HostSharedPtr, 16384> slots_;
+
+  // loop through the cluster slot response and error checks for each field
```

Comment:
nit: start comment sentences with capital, end with period, etc. Please check the entire change.